File: javascript/quickSort.js
function quickSort(arr, left = 0, right = arr.length - 1) {
    // Base case: if the current segment of the array is invalid or has only one element
    if (left >= right) {
        return;
    }

    // Partition the array and get the index of the pivot element after partition
    let pivotIndex = partition(arr, left, right);

    // Recursively apply quicksort to the segments before and after the partition
    quickSort(arr, left, pivotIndex - 1);
    quickSort(arr, pivotIndex + 1, right);

    // Return the sorted array
    return arr;
}

function partition(arr, left, right) {
    // Choose the rightmost element as the pivot
    const pivotValue = arr[right];
    // Starting index of the partition
    let partitionIndex = left;

    // Rearrange the array by comparing each element with the pivot
    for (let i = left; i < right; i++) {
        // If current element is smaller than the pivot, swap it to the correct position
        if (arr[i] < pivotValue) {
            [arr[i], arr[partitionIndex]] = [arr[partitionIndex], arr[i]];
            partitionIndex++;
        }
    }

    // Swap the pivot element to its final position
    [arr[partitionIndex], arr[right]] = [arr[right], arr[partitionIndex]];
    return partitionIndex; // Return the index of the pivot after partitioning
}

// Testing the quickSort function.
console.log(quickSort([121, 12, 132, 51, 61, 99, 200, 0, -1]))

File: python/quick_sort.py
def quick_sort(arr, left=0, right=None):
    if right is None:
        right = len(arr) - 1  # Initialize 'right' on the first call

    # Base condition: if the part of the array has only one element or is invalid
    if left >= right:
        return arr

    # Partition the array around a pivot and get the index of the pivot element
    pivot_index = partition(arr, left, right)

    # Recursively sort the parts of the array before and after the partition
    quick_sort(arr, left, pivot_index - 1)
    quick_sort(arr, pivot_index + 1, right)

    # Return the sorted array
    return arr

def partition(arr, left, right):
    # The rightmost element is chosen as the pivot
    pivot_value = arr[right]
    # The 'partition_index' will track the final position of the pivot
    partition_index = left

    for i in range(left, right):
        # If an element is smaller than the pivot, it is moved to the 'left' part
        if arr[i] < pivot_value:
            arr[i], arr[partition_index] = arr[partition_index], arr[i]  # Swap elements
            partition_index += 1  # Move the partition index to the right

    # Place the pivot at its final place (all elements left to it are smaller)
    arr[partition_index], arr[right] = arr[right], arr[partition_index]
    return partition_index  # Return the final position of the pivot


# Testing
if __name__ == "__main__":
    print(quick_sort([-1, -400, 121, 12, 132, 51, 61, 99]))  